core/src/main/scala/chisel3/reflect/DataMirror.scala

```diff
@@ -346,4 +349,23 @@ object DataMirror {
     }
     myItems ++ childItems
   }
+
+  // Function to find the least common ancestor of two nodes
+  private[chisel3] def leastCommonAncestorModule(left: HasId, right: HasId): Option[BaseModule] = {
+    val leftPathSet = modulePath(left, None).toSet
+    val rightPath = modulePath(right, None)
+    rightPath.collectFirst { case p if leftPathSet.contains(p) => p }
+  }
+  // Function to path upwards, stopping if reaching including
+  private[chisel3] def modulePath(h: HasId, until: Option[BaseModule]): Seq[BaseModule] = {
+    val me = h match {
+      case m: BaseModule => Seq(m)
+      case d: Data       => d.topBinding.location.toSeq
+      case m: MemBase[_] => m._parent.toSeq
+    }
+    if (me == until.toSeq) Nil
+    else {
+      me ++ me.flatMap(x => x._parent.toSeq.flatMap(p => modulePath(p, until)))
+    }
+  }
```

Comment:
This method is a bit weird with it's use of `Option` and casts to `Seq` only to `flatMap`. Also, this is using recursion, but is not tail recursive. This would likely be better if rewritten to use iteration.